storage/innobase/xtrabackup/src/xbcloud/xbcloud.h

```diff
@@ -0,0 +1,186 @@
+  }
+
+  /**
+  Get the next available file to be downloaded. The file is on file_list
+  (allocated to this thread) and has no in progress chunks been downloaded.
+
+  @param [in/out]  file  file available to download.
+
+  @return true if a file has been allocated, false otherwise */
+  bool next_file(file_metadata_t &file) {
+    std::lock_guard<std::mutex> g(m);
+    if (!file_list.empty()) {
+      for (auto it : file_list) {
+        if (in_progress_files.count(it.first) == 0) {
+          file = it.second;
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+  Get the next chunk for a given file.
+
+  @param [in]  file  file to get the next chunk.
+
+  @return id of next available chunk */
+  my_off_t next_chunk(file_metadata_t &file) {
+    std::lock_guard<std::mutex> g(m);
+    auto it = file_list.find(file.name);
+    assert(it != file_list.end());
+    return it->second.next_chunk;
+  }
+
+  /**
+  Attempt to reserve ownsership of a file with intend to start downloading it.
+  This will add the file into both in_progress_files list and file_list list.
+
+  @param [in]  file  file to attempt to reserve.
+
+  @return true if the file is not already in progress, false otherwise. */
+  bool start_file(const file_metadata_t &file) {
+    std::lock_guard<std::mutex> g(m);
+    if (in_progress_files.count(file.name) > 0) {
+      return false;
+    }
+    in_progress_files.insert(file.name);
+    file_list.insert({file.name, file});
+    return true;
+  }
+
+  /**
+  Indicates the current chunk of a given file has been completed.
+  This will remove the file from in_progress_files list. If this is the last
+  chunk for this file, it will remove the file from file_list otherwise advance
+  the next_chunk of the file.
+
+  @param [in]  file  filen to complete chunk.
+  @param [in]  idx   chunk index that we have completed. */
+  void complete_chunk(const file_metadata_t &file, const my_off_t &idx) {
+    std::lock_guard<std::mutex> g(m);
+    in_progress_files.erase(file.name);
+    auto it = file_list.find(file.name);
+    assert(it != file_list.end());
+    if (file.last_chunk == idx)
+      file_list.erase(it);
+    else
+      it->second.next_chunk++;
+  }
+};
+
+/* Thread safe global file list */
+struct file_list_t {
+  std::mutex m;
+  /* Global list of files to be downloaded*/
+  std::unordered_map<std::string, file_metadata_t> files;
+
+  /**
+  Check if the file list is empty.
+
+  @return false if there are files in the list, true otherwise. */
+  bool empty() {
+    std::lock_guard<std::mutex> g(m);
+    return files.empty();
+  }
+
+  /**
+  Get the next available file to be downloaded from global list. The file
+  will be removed from this list. The caller must allocate the file into
+  its own thread file list.
+
+  @param [in/out]  file  file available to download.
+
+  @return true if a file has been allocated, false otherwise */
+  bool next_file(file_metadata_t &file) {
```

Comment:
next_file() in two different contexts is confusing. WHen I see "next_file()" I should be able to make out if it is from global/per_thread.

Reply:
Done by renaming the object